File: 056.rb
# 056
# Merge Intervals
# 2015-12-11
#####################################################
# Definition for an interval.
class Interval
  attr_accessor :start, :end
  def initialize(s=0, e=0)
    @start = s
    @end = e
  end
  def to_s
    "[#{@start}, #{@end}]"
  end
end

# @param {Interval[]} intervals
# @return {Interval[]}
def merge(intervals)
  v = []
  return v if intervals.empty?
  intervals.sort!{|x,y| x.start <=> y.start}

  while intervals.size >= 2
    last = intervals.pop
    index = intervals.rindex{|x| last.start <= x.end}
    if index.nil?
      v.push(last)
    else
      lower = intervals[index].start
      upper = [intervals[index].end, last.end].max
      intervals.pop(intervals.size-index)
      intervals.push(Interval.new(lower,upper))
    end
  end
  v.push(intervals.first)
end

#####################################################

intervals = [
=begin
  Interval.new(2,6),
  Interval.new(1,3),
  Interval.new(8,10),
  Interval.new(15,18)
=end
=begin
  Interval.new(1,4),
  Interval.new(2,3)
=end
#=begin
  Interval.new(2,3),
  Interval.new(4,5),
  Interval.new(6,7),
  Interval.new(8,9),
  Interval.new(1,10)
#=end
=begin
  Interval.new(1,4),
  Interval.new(0,2),
  Interval.new(3,5)
=end

]

puts merge(intervals)

File: 056.py
#   056
#   Merge Intervals
#   2015-12-13
#####################################################
# Definition for an interval.
class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if not intervals:
            return []
        intervals.sort(key=lambda a: a.start)
        v = [intervals[0]]
        for a in intervals[1:]:
            if a.start <= v[-1].end:
                v[-1].end = max(v[-1].end, a.end)
            else:
                v.append(a)
        return v

#####################################################

intervals = [
    Interval(2,6),
    Interval(1,3),
    Interval(8,10),
    Interval(15,18)

    #Interval(1,4),
    #Interval(2,3)

    #Interval(2,3),
    #Interval(4,5),
    #Interval(6,7),
    #Interval(8,9),
    #Interval(1,10)

    #Interval(1,4),
    #Interval(0,2),
    #Interval(3,5)
]

sol = Solution()

for a in sol.merge(intervals):
    print a.start, a.end

